### ISSUE TITLE
UPS Mail Innovations

### ISSUE THREAD

**Issue by John-Nenniger** [jkeen/tracking_number_data#25]
I have a few number from UPS Mail Innovations which just started coming into my app and being labelled unknown

They are 18 digits, all numbers, and so far have this format 801184************ but it could just be a simple id counter (which seems likely) so the first few digits are likely irrelevant

According to this stack overflow answer, (https://stackoverflow.com/questions/38188373/how-to-distinguish-ups-mail-innovations-from-usps-tracking-number) they can't be distinguished from USPS, so this might be strictly impossible.   

The tracking number in the above question has 26 digits though, so the system seems to have changed since then.  Either that or there are multiple options:

```
Q. | What number do I use to track my mail piece?
A. | You can use the Package ID, Tracking number, or Delivery Confirmation Number.
```

I'll send a message to UPS and see if there are any consistencies that would make this doable.  Any advice is welcome

**Comment by jkeen** [comment]
I can't find any documentation on UPS Mail Innovation numbers anywhere, and this is really the first I've heard of them. 

The numbers you're getting aren't being recognized as USPS numbers?

**Comment by lastobelus** [comment]
@jkeen in the long run we post these to Shopify, and what we have observed is:

* whatever system shopify is using, it is able to get state info on the package
* the tracking number takes some time to be recognized by usps, perhaps a day. Until then ups just responds with nothing
* here's an example:
<img width="492" alt="Tracking__UPS_-_Canada" src="https://user-images.githubusercontent.com/25810/55593660-8a5d7b00-56f1-11e9-97f5-315d7d145876.png">

**Comment by lastobelus** [comment]
UPS doesn't have much info, but it seems like UPS mail innovations is some sort of collusion between UPS & USPS:

https://www.ups.com/us/en/help-center/sri/track-mi-package.page

**Comment by cookieguru** [comment]
TL;DR: https://postalpro.usps.com/pub199

UPS Mail Innovations barcodes will meet the USPS spec.  All of the following information also applies to USPS tracking numbers, FedEx SmartPost numbers, DHL GlobalMail, and any others that utilize the USPS network.

The full tracking number will start with 420 followed by the five digit zip code.  Why 420?  Because the spec says so (Pub 199 section 4.1 on page 48 of the January 19 2024 edition).  Those eight digits are never printed on the package.

Following this will be 92-95.  91 is no longer valid (Appendix J, Table 2).  Other prefixes are not valid either.  92 and 93 indicate a commercial mailer.  UPS MI numbers will always fall in these.  Following the 3 digit service type code will be the mailer ID.  This is the best way to identify whether or not a package is transported via UPS MI, FedEx SP, DHL GM, etc.  However the numbers do not correlate to the zone skipper but rather the sender.  For example mailer ID `999999837` is assigned to `AMAZON FS / DFW5` (Amazon Fulfilment Services) (former Woot) warehouse in Carrollton, TX.  `909900790` belongs to `FUJI FILM OR / UPS - MI` and all packages are shipped via UPS MI.  It's very common for each warehouse location to have its own mailer ID.  This is because tracking numbers are generated offline, and this prevents collisions.  `909815791` is a mailer ID used by Target, and I've seen this same mailer ID used for both FedEx SmartPost *and* UPS SurePost within a few days of each other.

There is nothing in the tracking number that can reliably be used to determine which zone skipper is responsible.  Shippers don't often change zone skippers, so with Target being the exception you can usually assume the same service would be used once you pinpoint the mailer IDs.

Note that Pub 199 not only describes the barcode format but also the data exchange format (think a USPS implementation of EDI) for billing.  Not all of the numbers are the same across the two usages.

**Comment by cookieguru** [comment]
> > The full tracking number will start with 420
> 
> Apparently not. I just came across the example of [UPS 92612999955380553026887812](https://www.ups.com/track?loc=en_US&tracknum=92612999955380553026887812&requester=ST/trackdetails), a "UPS Mail Innovations® Expedited" shipment from New Jersey to Texas.

You don't have the full tracking number.  The full tracking number of this shipment is `4201970192612999955380553026887812`.  And it's actually headed _to_ New Jersey _from_ Texas.

> it's not possible to distinguish a USPS-issued UPS Mail Innovations tracking number from an ordinary USPS tracking number
1. This is what I said more than a year ago
2. USPS isn't issuing the tracking numbers for UPS Mail Innovations

**Comment by jkeen** [comment]
@cookieguru @dan-jensen @szalkerous @lastobelus @John-Nenniger

Well well well look whose attention came back around to this issue—me! 

I've been trying some things in a local branch, and I agree that the root fix to this whole class of problems is what @cookieguru described above: changing the pattern definition to be a lookup on a USPS IMpb number definition to get the Shipper via the MailerID, instead of a definition on the shipper to match the carrier. USPS doesn't publish a Mailer ID registry, though, so the data can only grow from numbers somebody can name the carrier for. I've got FedEx SmartPost and DHL eCommerce data harvested from my own shipments and a public order export I found, but nothing for Mail Innovations or OSM Worldwide.

Do y'all still have any of those tracking numbers available to you that you can share? If you delete the last 8 digits and share the rest it would only name a mailer (and not a package or a person), and it would help out a lot.

```
92612903574522  DHL eCommerce
```

That works whether or not your numbers include the 420 + ZIP prefix.

**Comment by cookieguru** [comment]
@jkeen Did you actually read my message?  The mailer ID does **not** point to FedEx or DHL but rather the shipper themselves.  It is not possible to identify the zone skipper with the tracking number alone.

Also, your "just remove the last 8 digits" trick is not sufficient: a USPS tracking number can be 22, 26, 30, 34, or 38 digits long.

**Comment by jkeen** [comment]
@cookieguru Read but not comprehended, apparently! My bad, my local tests had me hopeful and deeply misguided apparently.

So I guess the move is to stop trying to match the original shipper from the number and identify all the IMpb numbers as USPS numbers with original shipper unknown. @lastobelus mentioned that they tried to run one of the numbers through USPS and it wasn't recognized… which has me confused about how that number operates as a tracking number for the recipient? I'm wondering if those numbers were just old and out of the system. 

Do you know if a recent and valid IMpb format UPS Mail Innovations tracking number is recognized on usps.com?

**Comment by cookieguru** [comment]
92023999991043543415313958

There's a huge difference in scans on usps.com vs ups.com but that's a different story

**Comment by jkeen** [comment]
Ok great, at least it scans, though. 

So identifying it as a USPS number with a USPS tracking link with maybe some additional info indicating that the original shipper can not be identified from the number alone? Does that sounds like the direction to you?

**Comment by cookieguru** [comment]
Read through Publication 199

---

The only thing you can say for sure with a tracking number:
* If the Channel Application Identifier is `92` or `93` then it _may_ be transported via a third party
* If it's 94 or 95 it won't be

Edit: the class of service can help narrow things down too, but isn't a guarantee

You can also get the shipping partner info via USPS' tracking API

**Comment by jkeen** [comment]
> You can also get the shipping partner info via USPS' tracking API

Hmm, interesting, didn't catch that on the tracking page the first time, but I see it now. Calling into carrier APIs is a line the `tracking_number` gem hasn't crossed and personally I like the boundary of being completely offline, decoding only, no tracking. But good to know, nonetheless

**Comment by jkeen** [comment]
Rebuilt those definitions based on that publication, and removed the conflicting definitions. [tracking_number_data v2.0](https://github.com/jkeen/tracking_number_data/releases/tag/v2.0.0), [tracking_number gem v3.0](https://github.com/jkeen/tracking_number/releases/tag/v3.0.0)

All the IMpb numbers are collapsed into two definitions that both resolve to USPS as the carrier. 

USPS IMpb N type: 
https://trackingnumber.fyi/420787459400111206206406260787

USPS IMpb C type:
https://trackingnumber.fyi/4201002334249200190132607600833457

**Comment by dan-jensen** [comment]
@jkeen outstanding work, as usual. Just upgraded 2 production apps from 2.4.0 to 3.0.0, with all tests green. Really appreciate you.
